### PR TITLE
7 detect instance variables in constructor

### LINKED ISSUE #7: detect instance variables in constructor
The current approach detects class variables (which is how dataclass works). But non-dataclass classes should be handled as well.

### LINKED ISSUE #13: Incorrect poetry link in PyPI project descripion
The link for Poetry in the Install section on https://pypi.org/project/py2puml/ goes to https://pipenv.readthedocs.io/en/latest/ rather than https://python-poetry.org/.

### PR DESCRIPTION
**Pull request by lucsorel**
closes #7, closes #13

### CODE REVIEW

**Comment by lucsorel** [comment]
hello @jjerphan @doyou89 @MMartin09

I am done with the implementation of a way to detect instance attributes in the body of a class constructor. Since you showed some interest for this feature, **I was wondering whether you would be interested in reviewing this pull-request?**

It is quite a big PR (I also split code and tests in different modules) and I would totally understand if you decline this suggestion  to involve the community in the evolution of this package.

By the way, I implemented that by parsing the abstract syntax tree of the constructor body, a little bit like @doyou89 suggested [there](https://github.com/lucsorel/py2puml/issues/7#issuecomment-867459416), but using the visitor pattern to handle:
- bnested

File: tests/py2puml/inspection/test_inspectclass.py
Comment (jjerphan) [reply tests/py2puml/inspection/test_inspectclass.py:59]: Also potentially, several attributes might not be detected, in that sense they could be reported in an assertion error at the end.

File: tests/py2puml/inspection/test_inspectclass.py
Comment (jjerphan) [inline tests/py2puml/inspection/test_inspectclass.py:58]: This part was not obvious at first read. Might this ease it?

```suggestion
        point_attribute: UmlAttribute = None
        for attribute in point_umlitem.attributes:
            if attribute.name == attribute_name
			    point_attribute = attribute
			    break
```

File: tests/py2puml/parsing/test_astvisitors.py
Comment (jjerphan) [inline tests/py2puml/parsing/test_astvisitors.py:25]: :+1:

**Review by jjerphan — CHANGES_REQUESTED** [review]
This is a super nice feature, thanks @lucsorel!

Here is a brief review based on tests (which all pass on my end).
I have not been into the implementations yet.

File: README.md
Comment (lucsorel) [reply README.md:12]: here, "business" does not mean "buying and selling" (_"les affaires"_), but is used in the "business domain" expression (_"le domaine métier"_), which applies for any Python project. I am not sure whether that was what you meant (business -> trading -> does not apply to research projects)? Moreover, what did you mean by "the application of your Python code"?

However, this sentence was particularly true when `py2puml` handled mainly data structures (dataclasses, namedtuples, enums). But now that it handles classes' constructors, it is also useful for other parts of an application (services, controllers, etc.).

I suggest some other formulations:

1. Generate PlantUML class diagrams to document your Python code.
2. Generate PlantUML class diagrams to document your Python application.
3. Generate PlantUML class diagrams to document the domain model of your Python application.
4. any other suggestions?

File: README.md
Comment (lucsorel) [reply README.md:33]: > Wouldn't it makes sense not to advocate for not versioning scripts

I think there is a double negation in this part but I got the meaning of the suggestion and I agree with the single-source-of-truth approach which would recommend to version code without what can be automatically derived from it.

However, here is what I came up doing in projects where I have no place to host some documentation:
1. I generate a PlantUML script with py2puml and version it so that new developers or project managers can view it online (in github / gitlab / bitbucket, which has a plugin to render it)
2. I ensure its consistency with a unit test which generates the class diagram with the Python API of py2puml and fails if the 2 PlantUML scripts do not match

Here is an example of such a "documentation unit-test": https://github.com/lucsorel/py2puml/blob/7-detect_instance_variables_in_constructor/tests/py2puml/test_py2puml.py#L10-L19

This is handy, but maybe not a practice to advocate :/

What do you think of this reformulation?
 `py2puml` outputs diagrams in PlantUML syntax, which can be hosted along other code documentation or maintained with a unit test (see [the `test_py2puml_model_on_py2uml_domain` test example](tests/py2puml/test_py2puml.py).).

File: README.md
Comment (lucsorel) [reply README.md:36]: agreed, I moved it at the bottom of the page, just before the `# alternatives` part.

File: README.md
Comment (lucsorel) [reply README.md:90]: my intention was to be able to make such a call:

```sh
py2puml /home/luc/repos/py2puml/py2puml/domain py2puml.domain
```

But it fails, for now.

I believe it is because of the way the Python import system works: the contents of `sys.path` and `PYTHONPATH` in such calls. When doing the aforementioned call, I believe that some logic should infer that /home/luc/repos/py2puml should be used as the root directory and add it somehow in the `sys.path` / `PYTHONPATH` in order to make the call work properly. Are you familiar with such things?

To answer your question: yes, py2puml py2puml/domain would (and should) be feasible. In another issue, this PR is already heavy :smiley:

File: py2puml/py2puml.domain.puml
Comment (lucsorel) [reply py2puml/py2puml.domain.puml:1]: this file is kept in sync with the code with this ["documentation" unit test](https://github.com/lucsorel/py2puml/blob/7-detect_instance_variables_in_constructor/tests/py2puml/test_py2puml.py#L10-L19), which fails if the file is not in sync with the state of the code (this test uses the py2puml Python API to produce the PlantUML script matching the code state).

File: tests/modules/withconstructor.py
Comment (lucsorel) [reply tests/modules/withconstructor.py:14]: nitpicking is ok :wink: 
My purpose was to define a class variable and was able only to think about PI (which is rather useless for the Point class... :shrug: ).

File: tests/modules/withconstructor.py
Comment (lucsorel) [reply tests/modules/withconstructor.py:25]: type annotations are necessary for `py2puml` to infer the type of the instance attributes while parsing the AST.

`py2puml` is able to infer the type of self.coordinates in the following 2 examples:

1. inferred from a typed parameter in the constructor signature:

```python
def __init__(self, coordinates: Coordinates):
  self.coordinates = coordinates
```

2. inferred from a typed local variable of the constructor body:

```python
def __init__(self, x, y):
  coordinates: Coordinates = Coordinates(x, y)
  self.coordinates = coordinates
```

Of course, it would be possible to infer the type in other cases like:
- assigning a primitive-typed value (`self.x = 0`, `self.unit = 'meter'`)
- assigning an instance of another class (`self.origin = Coordinates(0, 0)`)

But that would be for another release. In any case, I do not want to execute and evaluate code from the constructor. I may attempt to connect to a database, read or write some file, etc. This is my red line. And parsing AST is fun :stuck_out_tongue_closed_eyes:

File: tests/modules/withenum.py
Comment (lucsorel) [reply tests/modules/withenum.py:1]: yes, I find files easier to read that way (maybe because it just adds room between my IDE's menu bar and the first line of code). It may not be a proper habit or a recommended one :shrug:

File: tests/py2puml/inspection/test_inspectclass.py
Comment (lucsorel) [reply tests/py2puml/inspection/test_inspectclass.py:12]: I am afraid that such a file would grow to big with time. However, I could add a `test/asserts.__init__.py` where all these assertion functions could be imported from, wdyt?

File: tests/py2puml/inspection/test_inspectclass.py
Comment (lucsorel) [reply tests/py2puml/inspection/test_inspectclass.py:58]: ah, it is a question of habit and taste between for-break loops and comprehension/iterator. Actually, I kind of prefer iterators or comprehensions (I like these Python features): it is easy to forget to write the `break` instruction. Are you ok with keeping the iterator?

File: tests/py2puml/inspection/test_inspectclass.py
Comment (lucsorel) [reply tests/py2puml/inspection/test_inspectclass.py:59]: > Isn't the convention for assert's error message that it be assertive about the expected behavior?
Right, I fixed that into 'must be detected' . I changed that in other assertions too. Thanks.

> Also potentially, several attributes might not be detected, in that sense they could be reported in an assertion error at the end.
Yes, that's what I would probably do in some production code (like checking the consistency of some set of options, for example). But for unit tests, I prefer them to fail fast and write them in a concise manner (otherwise, I would need to create and populate a list of error messages, then raise an error if the list is not empty).

File: tests/py2puml/test_py2puml.py
Comment (lucsorel) [reply tests/py2puml/test_py2puml.py:19]: and it enables to maintain the PlantUML script up-to-date (the unit test would fail otherwise).

**Comment by lucsorel** [comment]
:pray:  thank you @jjerphan for your review. I included most of you suggestions in a single commit from my machine (some of your comments mentioning "it applies somewhere else too" - it did not feel consistent to commit from github and from my machine in a separate commit).

There are a couple discussions that are still open for your comments, answers or advice. I hope I don't take too much of your time, I am really happy to have someone reviewing "my own" contributions to this package.

**Comment by jjerphan** [comment]
> :pray: thank you @jjerphan for your review. I included most of you suggestions in a single commit from my machine (some of your comments mentioning "it applies somewhere else too" - it did not feel consistent to commit from github and from my machine in a separate commit).

No problems, the easiest for you the better. 

> There are a couple discussions that are still open for your comments, answers or advice. I hope I don't take too much of your time, I am really happy to have someone reviewing "my own" contributions to this package.

Yes, I started to read them. I am busy with other things currently but I'll come back to it once free.
I am glad to give feedback: to me, it looks like a relevant project which could get project maintainers interested.

And, pragmatically, it also helps me with my work. :slightly_smiling_face:

**Comment by lucsorel** [comment]
hello @jjerphan,

I am going to be less available for reviewing any comments for the next 3 weeks (holidays). I will be back the last week of August and am looking forward to merge this feature in September, the latest, if possible. Would that be ok for you?

**Comment by jjerphan** [comment]
Hi @lucsorel 

It's fine for me.

_Bonnes vacances!_ :beach_umbrella:

**Comment by lucsorel** [comment]
> It's fine for me.

I am not sure about what you meant:

1. you will have time to do some more review by the end of the month?
2. you have reviewed enough and find this PR ready to be merged?

**Comment by jjerphan** [comment]
I meant that I should have time to do more review in August.

File: README.md
Comment (jjerphan) [reply README.md:12]: > here, "business" does not mean "buying and selling" ("les affaires"), but is used in the "business domain" expression ("le domaine métier"), which applies for any Python project. I am not sure whether that was what you meant (business -> trading -> does not apply to research projects)? Moreover, what did you mean by "the application of your Python code"?

Thanks for clarifying. What I meant is that "_business domain_" (which refers to actual real world problems, be it commercial or not) is not that coined in communities dealing with abstract problems (e.g. some in research, scientific computing, ...). 

I find your second proposition more inclusive. :+1: 

Nevertheless, my original remark is secondary: it's mostly me nitpicking.

File: README.md
Comment (jjerphan) [reply README.md:32]: > I think there is a double negation in this part

Indeed, there's a spurious double negation.

I understand your use-case and can relate to it (online previews are useful). 

> What do you think of this reformulation?
> 
> > py2puml outputs diagrams in PlantUML syntax, which can be hosted along other code documentation or maintained with a unit test (see the test_py2puml_model_on_py2uml_domain test example.).
> 

I guess we can mention this setup. Ideally/IMO, the code-base should not be responsible for maintaining fixtures for easing documentation rendering and one must generate diagram _ad-hoc_ when needed for useful but non-essential use-cases for the project. 

Could [Brython](https://realpython.com/brython-python-in-browser/) help using `py2pulm` from a browser?

File: tests/modules/withconstructor.py
Comment (jjerphan) [reply tests/modules/withconstructor.py:26]: Thanks for the summary. I also think `py2pulm` should stick to parsing ASTs.

File: py2puml/inspection/inspectclass.py
Comment (jjerphan) [inline py2puml/inspection/inspectclass.py:81]:
```suggestion
            uml_attr = UmlAttribute(attr_name, attr_type, static=True)
```

File: tests/py2puml/inspection/test_inspectclass.py
Comment (jjerphan) [reply tests/py2puml/inspection/test_inspectclass.py:12]: > I am afraid that such a file would grow to big with time.

As of now, some files consist of solely a few lines. They can always be split then if growing too big.

Alternatively, I you want to stick to several files, adding a `__init__.py` for `tests.asserts` is probably suitable. This would also allowing avoiding long fully qualified module names.

File: tests/py2puml/inspection/test_inspectclass.py
Comment (jjerphan) [reply tests/py2puml/inspection/test_inspectclass.py:58]: Let's keep it as is to avoid adding nested scopes.

File: py2puml/inspection/inspectclass.py
Comment (jjerphan) [inline py2puml/inspection/inspectclass.py:117]: Why do you need to remove staticity here?

File: py2puml/inspection/inspectclass.py
Comment (jjerphan) [inline py2puml/inspection/inspectclass.py:88]: Does the `d` in `fqdn` stands for "domain"? Probably, using `m` for "module" is more correct here?

File: py2puml/parsing/astvisitors.py
Comment (jjerphan) [inline py2puml/parsing/astvisitors.py:120]: This suggestions also applies to other places in the code.
```suggestion
            self.uml_attributes.append(UmlAttribute(variable.id, short_type, static=False))
```

**Review by jjerphan — COMMENTED** [review]
Some remarks from my limited time. Feel free to integrate those changes (I am a sporadic user of this tool and would debug and/or add features then if I need them).

Some general suggestions:
 - apply black to format the code
 - merge files and only use a max depth of 2 for module nesting

File: py2puml/inspection/inspectclass.py
Comment (lucsorel) [reply py2puml/inspection/inspectclass.py:117]: because the @dataclass decorator creates a constructor and instance attributes from the static class attributes. Therefore, the static attribute definitions document the instance attributes, which are not static anymore.

File: py2puml/parsing/astvisitors.py
Comment (lucsorel) [reply py2puml/parsing/astvisitors.py:120]: thank you @jjerphan for your review, I will take a look at it this week.
Have a nice day

File: py2puml/parsing/astvisitors.py
Comment (doyou89) [inline py2puml/parsing/astvisitors.py:76]: This suggestions also applies to other places in the code.
```suggestion
self.class_self_id: str
```

**Review by doyou89 — COMMENTED** [review]
It has been going on for a long time. I will also participate in the review as soon as I have time.

File: py2puml/parsing/astvisitors.py
Comment (doyou89) [inline py2puml/parsing/astvisitors.py:98]:
```suggestion
    def get_from_namespace(self, variable_id: str) -> Optional[Variable]:
        return next((
            variable
            # variables namespace is iterated antichronologically
            # to account for variables being overridden
            for variable in self.variables_namespace[::-1]
            if variable.id == variable_id
        ), None)
```

File: py2puml/parsing/astvisitors.py
Comment (doyou89) [inline py2puml/parsing/astvisitors.py:149]:
```suggestion
            if (
                len(variables_collector.self_attributes) == 1
            ) and (
                isinstance(node.value, Name)
            ):
                assigned_variable = self.get_from_namespace(node.value.id)
                if assigned_variable:
                    short_type, full_namespaced_definitions = self.derive_type_annotation_details(
                        assigned_variable.type_expr
                    )
                    self.uml_attributes.append(
                        UmlAttribute(
                            variables_collector.self_attributes[0].id, short_type, False
                        )
                    )
                    self.extend_relations(full_namespaced_definitions)
```

**Review by doyou89 — COMMENTED** [review]
Suggest to remove the `is_in_variables_namespace` method and check directly where you use `get_from_namespace` method.

File: README.md
Comment (lucsorel) [reply README.md:12]: ok, I changed the wording for the 2nd suggestion

File: README.md
Comment (lucsorel) [reply README.md:32]: it is really nice discussing with you, I am learning new stuffs every time. I guess you suggested to use brighton.js to compute the class diagram and display it on the main README.md? Unfortunately, github prevents from hotlinking external css and js files.

I reworded the paragraph:

```text
`py2puml` outputs diagrams in PlantUML syntax, which can be:
* versioned along your code with a unit-test ensuring its consistency (see the [test_py2puml.py's test_py2puml_model_on_py2uml_domain](tests/py2puml/test_py2puml.py) example)
* generated and hosted along other code documentation (better option: generated documentation should not be versioned with the codebase)
```

File: py2puml/inspection/inspectclass.py
Comment (lucsorel) [reply py2puml/inspection/inspectclass.py:81]: agreed, it makes the source code more explicit. I changed that in other places too.

File: py2puml/inspection/inspectclass.py
Comment (lucsorel) [reply py2puml/inspection/inspectclass.py:88]: `fqdn` stands for "fully qualified domain name", a term that must be used for URLs. I got confused with the UML "fully qualified name" expression. I am fixing that too all over the codebase

File: py2puml/parsing/astvisitors.py
Comment (lucsorel) [reply py2puml/parsing/astvisitors.py:76]: I cannot change it in other places, it breaks the code. Here it is ok

File: py2puml/parsing/astvisitors.py
Comment (lucsorel) [reply py2puml/parsing/astvisitors.py:98]: agreed. There used to be several calls at some point, but there was just one now. Thanks

File: py2puml/parsing/astvisitors.py
Comment (lucsorel) [reply py2puml/parsing/astvisitors.py:120]: changed

File: py2puml/parsing/astvisitors.py
Comment (lucsorel) [reply py2puml/parsing/astvisitors.py:149]: agreed

File: tests/py2puml/inspection/test_inspectclass.py
Comment (lucsorel) [reply tests/py2puml/inspection/test_inspectclass.py:12]: I am fine with files with only a few lines of code, they keep the balance with other bigger files :smiley:

**Comment by lucsorel** [comment]
:pray: @jjerphan and @doyou89 for your reviews and insights. Thanks to you, py2puml@0.5.0 is out at https://pypi.org/project/py2puml/ :smiley:

**Pull request merged**
